File: 617.merge-two-binary-trees.java
/*
 * @lc app=leetcode id=617 lang=java
 *
 * [617] Merge Two Binary Trees
 */

// @lc code=start

import java.util.ArrayDeque;
import java.util.LinkedList;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public TreeNode mergeTrees(TreeNode root1, TreeNode root2) {
        if (root1 == null) return root2;
        if (root2 == null) return root1;
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root1); queue.add(root2);
        while (!queue.isEmpty()) {
            TreeNode n1 = queue.remove();
            TreeNode n2 = queue.remove();
            n1.val += n2.val;
            if (n1.left == null) n1.left = n2.left;
            else if (n1.left != null && n2.left != null) {
                queue.add(n1.left); queue.add(n2.left);
            }
            if (n1.right == null) n1.right = n2.right;
            else if (n1.right != null && n2.right != null) {
                queue.add(n1.right); queue.add(n2.right);
            }
        }
        return root1;
    }
}
// @lc code=end

File: 617.merge-two-binary-trees.cpp
/*
 * @lc app=leetcode id=617 lang=cpp
 *
 * [617] Merge Two Binary Trees
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <queue>
using namespace std;

class Solution {
public:
    TreeNode* mergeTrees(TreeNode* root1, TreeNode* root2) {
        if (root1 == nullptr) return root2;
        if (root2 == nullptr) return root1;
        queue<TreeNode *> qu;
        qu.push(root1); qu.push(root2);
        while (!qu.empty()) {
            TreeNode *n1 = qu.front(); qu.pop();
            TreeNode *n2 = qu.front(); qu.pop();
            n1->val += n2->val;
            if (n1->left != nullptr && n2->left != nullptr) {
                qu.push(n1->left); qu.push(n2->left);
            }
            if (n1->right != nullptr && n2->right != nullptr) {
                qu.push(n1->right); qu.push(n2->right);
            }
            if (n1->left == nullptr && n2->left != nullptr) {
                n1->left = n2->left;
            }
            if (n1->right == nullptr && n2->right != nullptr) {
                n1->right = n2->right;
            }
        }
        return root1;
    }
};
// @lc code=end
